src/scipp/curve_fit.py

```diff
@@ -516,37 +617,82 @@ def curve_fit(
         for arg, coord in coords.items()
     }
 
-    p0 = _make_defaults(f, coords.keys(), p0)
-
-    f = (
-        _wrap_scipp_func(f, p0)
-        if not unsafe_numpy_f
-        else _wrap_numpy_func(f, p0, coords.keys())
-    )
-
     map_over = tuple(
         d
         for d in da.dims
         if d not in reduce_dims and not any(d in c.dims for c in coords.values())
     )
 
-    # Create a dataarray with only the participating coords
-    _da = DataArray(da.data, coords=coords, masks=da.masks)
+    p0 = _make_defaults(f, coords.keys(), p0)
+    bounds = _parse_bounds(bounds, p0)
 
-    out = _prepare_numpy_outputs(da, p0, map_over)
+    pardim = None
+    if len(map_over) > 0:
+        fit_size = np.prod([da.sizes[dim] for dim in da.dims if dim not in map_over])
+        # Heuristic: if the fit size is above 10_000
+        # the multiprocessing likely does no good.
+        if workers is None and fit_size >= 10_000:
+            workers = 1
+
+        max_size = max((da.sizes[dim] for dim in map_over))
+        max_size_dim = next((dim for dim in map_over if da.sizes[dim] == max_size))
+        # Parallelize over longest dim because that is most likely
+        # to give us a balanced workload over the workers.
+        pardim = max_size_dim if max_size > 1 else None
+
+    # Only parallelize if the user did not explicitly ask for a single worker
+    # and a suitable dimension for parallelization was found.
+    if workers != 1 and pardim is not None:
```

Comment:
Relating to my comment in the tests regarding failing if the function cannot be pickled: This here demonstrates that this is indeed quite problematic. Say someone writes and tests some fitting code on small examples. Then `pardim` may bypass the pickling error. Then, when put into production we suddenly have `pardim != None` and the code that worked perfectly in the tests fails.

Can we make the behavior more consistent?